### PR TITLE
Feature/df 2690 idparc sso poc

### PR DESCRIPTION
**Pull request by palsch**

### CODE REVIEW

**Review by cjpsycos — APPROVED** [review]

**Review by birom — APPROVED** [review]

File: jobroom-gateway-app/src/test/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+import static org.mockito.Mockito.when;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+@AutoConfigureMockMvc
+@WithMockUserPrincipal
+public class IdparcSSOControllerTest {
```
Comment (seco-mgabor) [inline jobroom-gateway-app/src/test/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOControllerTest.java:37]: This is an integration test also. It should be renamed, to be able to run it in the integration test phase later on.
Usually @WebMvcTest is used to test MVC Controllers, but you should have your reasons why you did it like this. :)

File: jobroom-gateway-app/src/test/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+import java.util.Optional;
+
+import static ch.admin.seco.jobroom.web.rest.idparcsso.IdparcSSOUtils.IDPARC_SSO_JWT_COOKIE_NAME;
+import static ch.admin.seco.jobroom.web.rest.idparcsso.IdparcSSOUtils.IDPARC_SSO_JWT_NO_SAMESITE_COOKIE_NAME;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
```
Comment (seco-mgabor) [inline jobroom-gateway-app/src/test/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtilsTest.java:24]: Also an integration test.

**Review by seco-mgabor — CHANGES_REQUESTED** [review]
Please see my comments.

File: docker/gateway/central-server-config/docker-config/application.yml [hunk 3]
```diff
@@ -7,9 +7,20 @@ jhipster:
...
             defaultZone: http://admin:${jhipster.registry.password}@jhipster-registry:8761/eureka/
+
+app:
+    idparc-sso:
+        # base64 encoded "test-secret"
+        secret: dGVzdC1zZWNyZXQK
+        idParcLoginURL: 'https://onlineformulare.arbeit.swiss/'
+
+security:
+    user:
```
[the remark this answers is not in the document]
Comment (palsch) [reply docker/gateway/central-server-config/docker-config/application.yml:25]: To be consistent, I took this like it was used at other places for login-url

File: jobroom-gateway-app/src/main/resources/templates/idparc-sso/autosubmit-form.html [hunk 4]
```diff
@@ -0,0 +1,25 @@
+<!DOCTYPE html>
+<html xmlns:th="http://www.thymeleaf.org" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+      xsi:schemaLocation="http://www.thymeleaf.org ">
+<head>
+    <meta charset="utf-8" />
+</head>
+<body onload="document.forms[0].submit()">
+<noscript>
+    <p>
```
[the remark this answers is not in the document]
Comment (palsch) [reply jobroom-gateway-app/src/main/resources/templates/idparc-sso/autosubmit-form.html:9]: I wouldn't do that. If someone doesn't have JavaScript support, than he disabled it or uses a strange browser. So for this very specific case I wouldn't add the translation overhead.

[the remark this answers is not in the document]
Comment (seco-mgabor) [reply jobroom-gateway-app/src/main/resources/templates/idparc-sso/autosubmit-form.html:9]: If this is the case, it's OK for me too.

File: jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/config/SecurityConfiguration.java [hunk 5]
```diff
@@ -48,19 +44,22 @@
...
+import java.util.Arrays;
+import java.util.Collection;
+
+import static ch.admin.seco.jobroom.infrastructure.logging.BusinessLogEventType.USER_LOGOUT;
+import static ch.admin.seco.jobroom.infrastructure.logging.BusinessLogObjectType.USER;
+import static ch.admin.seco.jobroom.security.saml.infrastructure.dsl.SAMLConfigurer.saml;
+import static org.opensaml.saml2.core.AuthnContext.KERBEROS_AUTHN_CTX;
+import static org.opensaml.saml2.core.AuthnContext.NOMAD_TELEPHONY_AUTHN_CTX;
+import static org.opensaml.saml2.core.AuthnContext.SMARTCARD_PKI_AUTHN_CTX;
+import static org.opensaml.saml2.core.AuthnContext.SOFTWARE_PKI_AUTHN_CTX;
```
[the remark this answers is not in the document]
Comment (palsch) [reply jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/config/SecurityConfiguration.java:63]: I would rather not touch that file with my change. Maybe we could create a cleanup tech dept for this one.

File: jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/ProfileInfoResource.java [hunk 6]
```diff
@@ -45,9 +50,15 @@ private String getRibbonEnv(String[] activeProfiles) {
 
         private String ribbonEnv;
 
-        ProfileInfoVM(String[] activeProfiles, String ribbonEnv) {
+        private String idParcLoginURL;
+
```
[the remark this answers is not in the document]
Comment (palsch) [reply jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/ProfileInfoResource.java:54]: I dont know, it is not my code.

File: jobroom-gateway-app/src/test/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOControllerTest.java (hunk 1 above)
Comment (palsch) [reply jobroom-gateway-app/src/test/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOControllerTest.java:37]: All of the tests here are done in that way. So we would have to rename all of the files.
To be consistent, I used the same approach.

File: jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtils.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        response.addCookie(generateClearingCookie(IDPARC_SSO_JWT_NO_SAMESITE_COOKIE_NAME));
+    }
+
+    private Cookie generateClearingCookie(String name){
+        final Cookie c = new Cookie(name, null);
+        c.setMaxAge(0);
+        return c;
+    }
+
+    public void setNoCacheHeader(HttpServletResponse response) {
```
[the remark this answers is not in the document]
Comment (palsch) [reply jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtils.java:18]: To be safe, I would like to use both, see also https://developer.mozilla.org/de/docs/Web/HTTP/Headers/Cache-Control#Caching_verhindern

File: jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtils.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    public String extractIdparcJWTFromCookie(HttpServletRequest request) {
+        Cookie[] cookies = request.getCookies();
+        if (cookies == null) {
+            return null;
+        }
+
+        return Arrays.stream(cookies)
+            .filter(cookie -> cookie.getName().equals(IDPARC_SSO_JWT_COOKIE_NAME) ||
```
[the remark this answers is not in the document]
Comment (palsch) [reply jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtils.java]: I am not sure what you mean

File: jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtils.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        }
+        if (StringUtils.hasText(cookie.getDomain())) {
+            buf.append("; Domain=").append(cookie.getDomain());
+        }
+        int maxAge = cookie.getMaxAge();
+        if (maxAge >= 0) {
+            buf.append("; Max-Age=").append(maxAge);
+            buf.append("; Expires=");
+            org.springframework.http.HttpHeaders headers = new org.springframework.http.HttpHeaders();
+            headers.setExpires(maxAge > 0 ? System.currentTimeMillis() + 1000L * maxAge : 0);
```
[the remark this answers is not in the document]
Comment (palsch) [reply jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtils.java]: It's like ready to use date calculation. So lesser code, no need to test.

File: jobroom-gateway-app/src/test/java/ch/admin/seco/jobroom/application/idparcsso/IdparcSSOServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+import static ch.admin.seco.jobroom.security.AuthoritiesConstants.ROLE_COMPANY;
+import static ch.admin.seco.jobroom.security.AuthoritiesConstants.ROLE_JOBSEEKER_CLIENT;
+import static ch.admin.seco.jobroom.security.AuthoritiesConstants.ROLE_PRIVATE_EMPLOYMENT_AGENT;
+import static io.jsonwebtoken.Jwts.parser;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
```
[the remark this answers is not in the document]
Comment (palsch) [reply jobroom-gateway-app/src/test/java/ch/admin/seco/jobroom/application/idparcsso/IdparcSSOServiceTest.java:38]: All of the tests here are done in that way. So we would have to rename all of the files.
To be consistent, I used the same approach.

File: jobroom-gateway-app/src/test/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtilsTest.java (hunk 2 above)
Comment (palsch) [reply jobroom-gateway-app/src/test/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtilsTest.java:13]: All of the tests here are done in that way. So we would have to rename all of the files.
To be consistent, I used the same approach.

File: jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtils.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+     *
+     * <p>see also <a href="https://caniuse.com/#feat=same-site-cookie-attribute">SameSite browser support</a>
+     */
+    public void saveIdparcJwtToCookie(String jwtToken, HttpServletResponse response) {
+        String cookieNoSameSite = generateIdparcJwtCookie(IDPARC_SSO_JWT_NO_SAMESITE_COOKIE_NAME, jwtToken, null);
+        String cookieWithSameSite = generateIdparcJwtCookie(IDPARC_SSO_JWT_COOKIE_NAME, jwtToken, "None");
+        response.addHeader(HttpHeaders.SET_COOKIE, cookieNoSameSite);
+        response.addHeader(HttpHeaders.SET_COOKIE, cookieWithSameSite);
+    }
+
```
[the remark this answers is not in the document]
Comment (palsch) [reply jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtils.java]: done

File: jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtils.java [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+        }
+
+        return Arrays.stream(cookies)
+            .filter(cookie -> cookie.getName().equals(IDPARC_SSO_JWT_COOKIE_NAME) ||
+                cookie.getName().equals(IDPARC_SSO_JWT_NO_SAMESITE_COOKIE_NAME))
+            .findFirst()
+            .map(Cookie::getValue)
+            .orElse(null);
+    }
+
```
[the remark this answers is not in the document]
Comment (palsch) [reply jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtils.java]: done

File: jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtils.java [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+        c.setMaxAge(0);
+        return c;
+    }
+
+    public void setNoCacheHeader(HttpServletResponse response) {
+        response.addHeader("Cache-Control", "no-cache, no-store");
+    }
+
+    public String generateAutosubmitForm(String actionUrl, String ssoJwt) {
+        Context context = new Context();
```
[the remark this answers is not in the document]
Comment (palsch) [reply jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtils.java:30]: done

File: jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtils.java [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+    private String generateIdparcJwtCookie(String name, String jwtToken, String sameSite) {
+        final Cookie c = new Cookie(name, jwtToken);
+        c.setSecure(idParcProperties.isUseSecureCookie());
+        c.setPath("/");
+        c.setHttpOnly(true);
+        c.setMaxAge(idParcProperties.getCookieMaxAge());
+        return getCookieHeaderString(c, sameSite);
+    }
+
```
[the remark this answers is not in the document]
Comment (palsch) [reply jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtils.java]: done

File: jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtils.java [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+    /**
+     * Saves two cookies with same value as workaround because of different "SameSite=None" browser behaviors (e.g. Safari).
+     * <p>The browser has to send us the cookie also in that case, when IAM (different domain) is doing the request to our service.
+     * This is only possible, if the SameSite attribute is set to "None".
+     * Safari has a bug, that ignores such cookies. So we use two cookies, with and without SameSite attribute.</p>
+     *
+     * <p>see also <a href="https://caniuse.com/#feat=same-site-cookie-attribute">SameSite browser support</a>
+     */
```
[the remark this answers is not in the document]
Comment (palsch) [reply jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtils.java]: done

File: jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOController.java [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+import javax.servlet.http.HttpServletResponse;
+
+import static ch.admin.seco.jobroom.infrastructure.logging.BusinessLogEventType.IDPARC_SSO_ATTEMPT;
+import static ch.admin.seco.jobroom.infrastructure.logging.BusinessLogEventType.IDPARC_SSO_SUCCESSFUL;
+import static ch.admin.seco.jobroom.infrastructure.logging.BusinessLogEventType.IDPARC_SSO_USER_NOT_REGISTERED;
+
+@RestController
+@RequestMapping("/api/idparc-sso")
+public class IdparcSSOController {
+
```
[the remark this answers is not in the document]
Comment (palsch) [reply jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOController.java:37]: done

File: jobroom-gateway-app/src/test/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtilsTest.java (hunk 2 above)
Comment (seco-mgabor) [reply jobroom-gateway-app/src/test/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtilsTest.java:13]: We will rename them. This is a new test, therefore I would have preffered to be named correctly. Less work in the future...

File: jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtils.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (seco-mgabor) [reply jobroom-gateway-app/src/main/java/ch/admin/seco/jobroom/web/rest/idparcsso/IdparcSSOUtils.java]: Text was hidden :(

**Review by seco-mgabor — APPROVED** [review]

**Pull request merged**